File: astro/rulers.py
"""Управители знаков, достоинства и цепочки диспозиторов."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Mapping, Optional, Sequence, Tuple

from .zodiac import SIGNS

TRADITIONAL = "traditional"
MODERN = "modern"

#: Управители знаков по порядку от Овна до Рыб.
TRADITIONAL_RULERS: Tuple[str, ...] = (
    "mars", "venus", "mercury", "moon", "sun", "mercury",
    "venus", "mars", "jupiter", "saturn", "saturn", "jupiter",
)

#: Современная схема: высшие планеты управляют Скорпионом, Водолеем и Рыбами.
MODERN_RULERS: Tuple[str, ...] = (
    "mars", "venus", "mercury", "moon", "sun", "mercury",
    "venus", "pluto", "jupiter", "saturn", "uranus", "neptune",
)

#: Классический управитель остаётся вторым для знаков с высшим управителем.
CO_RULERS: Dict[int, str] = {7: "mars", 10: "saturn", 11: "jupiter"}

SCHEMES = {TRADITIONAL: TRADITIONAL_RULERS, MODERN: MODERN_RULERS}

#: Экзальтации: тело → (индекс знака, градус наибольшей силы).
EXALTATIONS: Dict[str, Tuple[int, float]] = {
    "sun": (0, 19.0),       # Овен
    "moon": (1, 3.0),       # Телец
    "mercury": (5, 15.0),   # Дева
    "venus": (11, 27.0),    # Рыбы
    "mars": (9, 28.0),      # Козерог
    "jupiter": (3, 15.0),   # Рак
    "saturn": (6, 21.0),    # Весы
}

DIGNITY_NAMES = {
    "domicile": "обитель",
    "exaltation": "экзальтация",
    "detriment": "изгнание",
    "fall": "падение",
    "peregrine": "перегрин",
}


def rulers(scheme: str = TRADITIONAL) -> Tuple[str, ...]:
    """Таблица управителей выбранной схемы."""
    try:
        return SCHEMES[scheme]
    except KeyError:
        raise ValueError(f"неизвестная схема управителей: {scheme!r}") from None


def ruler_of(sign_index: int, scheme: str = TRADITIONAL) -> str:
    """Управитель знака."""
    return rulers(scheme)[sign_index % 12]


def co_ruler_of(sign_index: int, scheme: str = TRADITIONAL) -> Optional[str]:
    """Второй управитель знака — классический, если основной высший."""
    if scheme != MODERN:
        return None
    return CO_RULERS.get(sign_index % 12)


def signs_ruled_by(body: str, scheme: str = TRADITIONAL) -> Tuple[int, ...]:
    """Индексы знаков, которыми управляет тело."""
    return tuple(i for i, r in enumerate(rulers(scheme)) if r == body)


def dignity(body: str, sign_index: int, scheme: str = TRADITIONAL) -> str:
    """Эссенциальное достоинство тела в знаке.

    Проверяются четыре классических состояния; если ни одно не подходит,
    тело перегрин — без достоинства и без ущерба.
    """
    sign_index %= 12
    table = rulers(scheme)
    if table[sign_index] == body:
        return "domicile"
    if table[(sign_index + 6) % 12] == body:
        return "detriment"
    exaltation = EXALTATIONS.get(body)
    if exaltation is not None:
        if exaltation[0] == sign_index:
            return "exaltation"
        if (exaltation[0] + 6) % 12 == sign_index:
            return "fall"
    return "peregrine"


def dignity_name(code: str) -> str:
    return DIGNITY_NAMES.get(code, code)


@dataclass(frozen=True)
class DispositorTree:
    """Результат разбора цепочек диспозиторов.

    ``dispositor`` — непосредственный диспозитор каждого тела: управитель
    знака, в котором тело стоит. ``chains`` — путь от тела до входа в
    замкнутый круг. ``cycles`` — сами круги: цикл длиной один означает
    планету в своей обители, то есть финального диспозитора карты; цикл
    длиннее одного — взаимная рецепция или более длинное кольцо, у карты
    тогда единого финального диспозитора нет.
    """

    dispositor: Mapping[str, Optional[str]]
    chains: Mapping[str, Tuple[str, ...]]
    cycles: Tuple[Tuple[str, ...], ...]
    terminal: Mapping[str, Optional[Tuple[str, ...]]]

    @property
    def final_dispositors(self) -> Tuple[str, ...]:
        """Планеты в своей обители — конечные точки всех цепочек."""
        return tuple(cycle[0] for cycle in self.cycles if len(cycle) == 1)

    @property
    def mutual_receptions(self) -> Tuple[Tuple[str, str], ...]:
        """Пары планет, взаимно управляющих знаками друг друга."""
        return tuple(
            (cycle[0], cycle[1]) for cycle in self.cycles if len(cycle) == 2
        )

    def chain_of(self, body: str) -> Tuple[str, ...]:
        return self.chains.get(body, ())


def build_dispositors(
    sign_by_body: Mapping[str, int],
    scheme: str = TRADITIONAL,
    bodies: Optional[Sequence[str]] = None,
) -> DispositorTree:
    """Строит диспозиторы для тел карты.

    ``sign_by_body`` — ключ тела → индекс знака, в котором оно стоит.
    Учитываются только те управители, которые сами есть в карте: если
    схема традиционная, а Плутон в карте есть, диспозитором он не станет,
    потому что знаками не управляет.
    """
    present = set(bodies) if bodies is not None else set(sign_by_body)

    dispositor: Dict[str, Optional[str]] = {}
    for body in sign_by_body:
        if body not in present:
            continue
        ruler = ruler_of(sign_by_body[body], scheme)
        dispositor[body] = ruler if ruler in present and ruler in sign_by_body else None

    chains: Dict[str, Tuple[str, ...]] = {}
    terminal: Dict[str, Optional[Tuple[str, ...]]] = {}
    cycles: List[Tuple[str, ...]] = []
    cycle_by_member: Dict[str, Tuple[str, ...]] = {}

    for start in dispositor:
        path: List[str] = []
        index_of: Dict[str, int] = {}
        node: Optional[str] = start
        while node is not None and node not in index_of:
            index_of[node] = len(path)
            path.append(node)
            node = dispositor.get(node)

        if node is None:  # цепочка оборвалась: управителя нет в карте
            chains[start] = tuple(path[1:])
            terminal[start] = None
            continue

        if node in cycle_by_member:  # пришли в уже известный круг
            cycle = cycle_by_member[node]
        else:
            cycle = tuple(path[index_of[node]:])
            # круг записывается всегда с одного и того же тела, чтобы
            # один и тот же цикл не появлялся в разных поворотах
            pivot = cycle.index(min(cycle))
            cycle = cycle[pivot:] + cycle[:pivot]
            cycles.append(cycle)
            for member in cycle:
                cycle_by_member[member] = cycle

        chains[start] = tuple(path[: index_of[node]][1:])
        terminal[start] = cycle

    cycles.sort(key=lambda c: (len(c), c))
    return DispositorTree(dispositor, chains, tuple(cycles), terminal)


def describe_chain(
    body: str,
    tree: DispositorTree,
    names: Optional[Mapping[str, str]] = None,
) -> str:
    """Человекочитаемая цепочка диспозиторов: от тела до кольца.

    ``names`` подменяет ключи тел на любые подписи, например русские.
    """
    def label(key: str) -> str:
        return names.get(key, key) if names else key

    steps: List[str] = [label(body)]
    steps.extend(label(k) for k in tree.chain_of(body))
    cycle = tree.terminal.get(body)
    if cycle is None:
        return " \u2192 ".join(steps) + " \u2192 ?"
    if len(cycle) == 1:
        if len(steps) == 1 and cycle[0] == body:
            return f"{label(body)} в своей обители"
        steps.append(label(cycle[0]))
        return " \u2192 ".join(steps) + " (финальный диспозитор)"
    ring_parts = [label(k) for k in cycle]
    ring = (" \u21c4 ".join(ring_parts) if len(cycle) == 2
            else " \u2192 ".join(ring_parts + ring_parts[:1]))
    if body in cycle:
        return f"{label(body)} в кольце: {ring}"
    return " \u2192 ".join(steps) + f" \u2192 кольцо: {ring}"

File: web/astro/rulers.js
// Управители знаков, достоинства и цепочки диспозиторов.

import { signIndex } from './zodiac.js';

export const TRADITIONAL = 'traditional';
export const MODERN = 'modern';

const TRADITIONAL_RULERS = [
  'mars', 'venus', 'mercury', 'moon', 'sun', 'mercury',
  'venus', 'mars', 'jupiter', 'saturn', 'saturn', 'jupiter',
];

const MODERN_RULERS = [
  'mars', 'venus', 'mercury', 'moon', 'sun', 'mercury',
  'venus', 'pluto', 'jupiter', 'saturn', 'uranus', 'neptune',
];

const SCHEMES = { [TRADITIONAL]: TRADITIONAL_RULERS, [MODERN]: MODERN_RULERS };

// Экзальтации: тело — знак и градус наибольшей силы.
export const EXALTATIONS = {
  sun: [0, 19], moon: [1, 3], mercury: [5, 15], venus: [11, 27],
  mars: [9, 28], jupiter: [3, 15], saturn: [6, 21],
};

export const DIGNITY_NAMES = {
  domicile: 'обитель', exaltation: 'экзальтация', detriment: 'изгнание',
  fall: 'падение', peregrine: 'перегрин', triplicity: 'триплицитет',
  term: 'терм', decan: 'декан',
};

// Триплицитеты по Дорофею: стихия — дневной, ночной и участвующий управитель.
const TRIPLICITY = [
  ['sun', 'jupiter', 'saturn'],
  ['venus', 'moon', 'mars'],
  ['saturn', 'mercury', 'jupiter'],
  ['venus', 'mars', 'moon'],
];

// Египетские термы: знак — управитель и верхняя граница в градусах знака.
const TERMS = [
  [['jupiter', 6], ['venus', 12], ['mercury', 20], ['mars', 25], ['saturn', 30]],
  [['venus', 8], ['mercury', 14], ['jupiter', 22], ['saturn', 27], ['mars', 30]],
  [['mercury', 6], ['jupiter', 12], ['venus', 17], ['mars', 24], ['saturn', 30]],
  [['mars', 7], ['venus', 13], ['mercury', 19], ['jupiter', 26], ['saturn', 30]],
  [['jupiter', 6], ['venus', 11], ['saturn', 18], ['mercury', 24], ['mars', 30]],
  [['mercury', 7], ['venus', 17], ['jupiter', 21], ['mars', 28], ['saturn', 30]],
  [['saturn', 6], ['mercury', 14], ['jupiter', 21], ['venus', 28], ['mars', 30]],
  [['mars', 7], ['venus', 11], ['mercury', 19], ['jupiter', 24], ['saturn', 30]],
  [['jupiter', 12], ['venus', 17], ['mercury', 21], ['saturn', 26], ['mars', 30]],
  [['mercury', 7], ['jupiter', 14], ['venus', 22], ['saturn', 26], ['mars', 30]],
  [['mercury', 7], ['venus', 13], ['jupiter', 20], ['mars', 25], ['saturn', 30]],
  [['venus', 12], ['jupiter', 16], ['mercury', 19], ['mars', 28], ['saturn', 30]],
];

// Халдейский ряд планет по убыванию видимой скорости.
const CHALDEAN = ['saturn', 'jupiter', 'mars', 'sun', 'venus', 'mercury', 'moon'];

export function rulerOf(sign, scheme = TRADITIONAL) {
  return SCHEMES[scheme][((sign % 12) + 12) % 12];
}

export function triplicityRuler(sign, diurnal) {
  const [day, night] = TRIPLICITY[(((sign % 12) + 12) % 12) % 4];
  return diurnal ? day : night;
}

export function termRuler(longitude) {
  const sign = signIndex(longitude);
  const degree = ((longitude % 30) + 30) % 30;
  for (const [ruler, bound] of TERMS[sign]) {
    if (degree < bound) return ruler;
  }
  return TERMS[sign][TERMS[sign].length - 1][0];
}

// Деканы по халдейскому ряду: первый декан Овна достаётся Марсу.
export function decanRuler(longitude) {
  const sign = signIndex(longitude);
  const decan = Math.floor((((longitude % 30) + 30) % 30) / 10);
  return CHALDEAN[(CHALDEAN.indexOf('mars') + sign * 3 + decan) % 7];
}

export function exaltationRulerOf(sign) {
  const index = ((sign % 12) + 12) % 12;
  for (const [body, [exaltSign]] of Object.entries(EXALTATIONS)) {
    if (exaltSign === index) return body;
  }
  return null;
}

export function majorState(body, sign, scheme = TRADITIONAL) {
  const index = ((sign % 12) + 12) % 12;
  const table = SCHEMES[scheme];
  if (table[index] === body) return 'domicile';
  if (table[(index + 6) % 12] === body) return 'detriment';
  const exaltation = EXALTATIONS[body];
  if (exaltation) {
    if (exaltation[0] === index) return 'exaltation';
    if ((exaltation[0] + 6) % 12 === index) return 'fall';
  }
  return 'peregrine';
}

// Все пять достоинств сразу. Важное различие: тело без обители и
// экзальтации ещё не перегрин — у него могут быть терм или декан.
// Перегрин — тот, у кого нет ни одного из пяти.
export function essentialDignities(body, longitude, diurnal, scheme = TRADITIONAL) {
  const sign = signIndex(longitude);
  const ruler = rulerOf(sign, scheme);
  const exaltation = exaltationRulerOf(sign);
  const triplicity = triplicityRuler(sign, diurnal);
  const term = termRuler(longitude);
  const decan = decanRuler(longitude);
  const own = [];
  if (ruler === body) own.push('domicile');
  if (exaltation === body) own.push('exaltation');
  if (triplicity === body) own.push('triplicity');
  if (term === body) own.push('term');
  if (decan === body) own.push('decan');
  return {
    body, sign, ruler, exaltation, triplicity, term, decan, own,
    state: majorState(body, sign, scheme),
    peregrine: own.length === 0,
  };
}

// Диспозиторы: у каждого тела — управитель его знака. Цепочки сходятся в
// кольца; кольцо из одного тела означает планету в своей обители, то есть
// финального диспозитора, из двух — взаимную рецепцию.
export function buildDispositors(signByBody, scheme = TRADITIONAL) {
  const present = new Set(Object.keys(signByBody));
  const dispositor = new Map();
  for (const [body, sign] of Object.entries(signByBody)) {
    const ruler = rulerOf(sign, scheme);
    dispositor.set(body, present.has(ruler) ? ruler : null);
  }

  const chains = new Map();
  const terminal = new Map();
  const cycles = [];
  const cycleByMember = new Map();

  for (const start of dispositor.keys()) {
    const path = [];
    const seen = new Map();
    let node = start;
    while (node !== null && node !== undefined && !seen.has(node)) {
      seen.set(node, path.length);
      path.push(node);
      node = dispositor.get(node);
    }
    if (node === null || node === undefined) {
      chains.set(start, path.slice(1));
      terminal.set(start, null);
      continue;
    }
    let cycle = cycleByMember.get(node);
    if (!cycle) {
      cycle = path.slice(seen.get(node));
      const pivot = cycle.indexOf([...cycle].sort()[0]);
      cycle = cycle.slice(pivot).concat(cycle.slice(0, pivot));
      cycles.push(cycle);
      cycle.forEach((member) => cycleByMember.set(member, cycle));
    }
    chains.set(start, path.slice(1, seen.get(node)));
    terminal.set(start, cycle);
  }

  cycles.sort((a, b) => a.length - b.length);
  return {
    dispositor,
    chains,
    terminal,
    cycles,
    finalDispositors: cycles.filter((c) => c.length === 1).map((c) => c[0]),
    mutualReceptions: cycles.filter((c) => c.length === 2),
  };
}
